numpy/lib/function_base.py

```diff
@@ -356,7 +356,7 @@ def jhat(nbins):
         p_k = np.histogram(x, bins=nbins)[0] / n
         return (2 - (n + 1) * p_k.dot(p_k)) / hh
 
-    nbins = min(range(1, max(n // 4, 2)), key=jhat)
+    nbins = min(range(1, max(100, int(np.sqrt(n)))), key=jhat)
```

Comment:
There is no upper bound in general, so a search with a fixed upper limit will not always work.